// unstorage/drivers/http.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;
var _utils = require("./utils/index.cjs");
var _ofetch = require("ofetch");
var _ufo = require("ufo");
var _default = (0, _utils.defineDriver)((opts = {}) => {
  const r = (key = "") => (0, _ufo.joinURL)(opts.base, key.replace(/:/g, "/"));
  const rBase = (key = "") => (0, _ufo.joinURL)(opts.base, (key || "/").replace(/:/g, "/"), ":");
  return {
    hasItem(key) {
      return (0, _ofetch.$fetch)(r(key), {
        method: "HEAD"
      }).then(() => true).catch(() => false);
    },
    async getItem(key) {
      const value = await (0, _ofetch.$fetch)(r(key));
      return value;
    },
    async getItemRaw(key) {
      const value = await (0, _ofetch.$fetch)(r(key), {
        headers: {
          accept: "application/octet-stream"
        }
      });
      return value;
    },
    async getMeta(key) {
      const res = await _ofetch.$fetch.raw(r(key), {
        method: "HEAD"
      });
      let mtime = void 0;
      const _lastModified = res.headers.get("last-modified");
      if (_lastModified) {
        mtime = new Date(_lastModified);
      }
      return {
        status: res.status,
        mtime
      };
    },
    async setItem(key, value) {
      await (0, _ofetch.$fetch)(r(key), {
        method: "PUT",
        body: value
      });
    },
    async setItemRaw(key, value) {
      await (0, _ofetch.$fetch)(r(key), {
        method: "PUT",
        body: value,
        headers: {
          "content-type": "application/octet-stream"
        }
      });
    },
    async removeItem(key) {
      await (0, _ofetch.$fetch)(r(key), {
        method: "DELETE"
      });
    },
    async getKeys(base) {
      const value = await (0, _ofetch.$fetch)(rBase(base));
      return Array.isArray(value) ? value : [];
    },
    async clear(base) {
      await (0, _ofetch.$fetch)(rBase(base), {
        method: "DELETE"
      });
    }
  };
});
module.exports = _default;
